README.md
# NAME

Audio::TinySoundFont - Interface to TinySoundFont, a "SoundFont2 synthesizer library in a single C/C++ file"

# SYNOPSIS

    use Audio::TinySoundFont;
    my $tsf = Audio::TinySoundFont->new('soundfont.sf2');
    $tsf->note_on( 'Clarinet', 60, 1.0 );
    my $samples = $tsf->render( seconds => 5 );
    my $preset = $tsf->preset('Clarinet');
    $tsf->note_off( $preset, 60 );
    $samples .= $tsf->render( seconds => 5 );
    
    # Using the Preset object
    my $preset = $tsf->preset('Clarinet');
    my $samples = $preset->render(
      seconds => 5,
      note => 60,
      vel => 0.7,
      volume => .3,
    );
    
    # Using the Builder object
    my $builder = $tsf->new_builder(
      [
        {
          preset => 'Clarinet',
          note   => 59,
          at     => 0,
          for    => 2,
        },
      ]
    );
    $builder->add(
      [
        {
          preset     => $preset,
          note       => 60,
          at         => 44100,
          for        => 44100 * 2,
          in_seconds => 0,
        },
      ]
    );
    my $samples = $builder->render;

# DESCRIPTION

Audio::TinySoundFont is a wrapper around [TinySoundFont](https://github.com/schellingb/TinySoundFont),
a "SoundFont2 synthesizer library in a single C/C++ file". This allows you to
load a SoundFont file and synthesize samples from it.

# CONSTRUCTOR

## new

    my $tsf = Audio::TinySoundFont->new($soundfont_file, %attributes)

Construct a new Audio::TinySoundFont object using the provided `$soundfont_file`
file. It can be a filename, a file handle, or a string reference to the contents
of a SoundFont.

### Attributes

- volume

    Set the initial, global volume. This is a floating point number between 0.0 and
    1.0. The higher the number, the louder the output samples.

# METHODS

## volume

    my $volume = $tsf->volume;
    $tsf->volume( 0.5 );

Get or set the current global volume. This is a floating point number between
0.0 and 1.0. The higher the number, the louder the output samples.

## preset\_count

    my $count = $tsf->preset_count;

The number of presets available in the SoundFont.

## presets

    my %presets = %{ $tsf->presets }

A HashRef of all of the presets in the SoundFont. The key is the name of the
preset and the value is the [Audio::TinySoundFont::Preset](https://metacpan.org/pod/Audio%3A%3ATinySoundFont%3A%3APreset) object for that
preset.

## preset

    my $preset = $tsf->preset('Clarinet');

Get a [Audio::TinySoundFont::Preset](https://metacpan.org/pod/Audio%3A%3ATinySoundFont%3A%3APreset) object by name. This will croak if the
preset name is not found.

## preset\_index

    my $preset = $tsf->preset_index($index);

Get an [Audio::TinySoundFont::Preset](https://metacpan.org/pod/Audio%3A%3ATinySoundFont%3A%3APreset) object by index in the SoundFont. This
will croak if the index is out of range. Note, this will return a different
object than ["preset"](#preset) will, which will return the object from [presets](https://metacpan.org/pod/presets).

## SAMPLE\_RATE

    my $sample_rate = $tsf->SAMPLE_RATE

Returns the sample rate that TinySoundFont is operating on, expressed as hertz
or samples per second. This is currently static and is 44\_100;

## new\_builder

    my $builder = $tsf->new_builder

Create a new [Audio::TinySoundFont::Builder](https://metacpan.org/pod/Audio%3A%3ATinySoundFont%3A%3ABuilder) object. This can be used to
generate a single sample from a script of what notes to play when.

## active\_voices

    my $count = $tsf->active_voices;

Returns the number of currently active voices that TinySoundFont is rendering.
Generally speaking, each ["note\_on"](#note_on) will make one or more voices active.

## is\_active

    my $bool = $tsf->is_active

Returns if TinySoundFont currently has active voices and will output audio
during render.

## note\_on

    $tsf->note_on($preset, $note, $velocity);

Turns a note on for a Preset. `$preset` can either be a
[Audio::TinySoundFont::Preset](https://metacpan.org/pod/Audio%3A%3ATinySoundFont%3A%3APreset) object or the name of a Preset. Both `$note`
and `$velocity` are optional. `$note` is a MIDI note between 0 and 127, with
60 being middle C and the default if it is not given. `$velocity` is a floating
point number between 0.0 and 1.0 with the default being 0.5.

## note\_off

    $tsf->note_off($preset, $note);

Turns a note off for a Preset. `$preset` can either be a
[Audio::TinySoundFont::Preset](https://metacpan.org/pod/Audio%3A%3ATinySoundFont%3A%3APreset) object or the name of a Preset. `$note` is
optional and is the same MIDI note given on ["note\_on"](#note_on). The default is 60.
This will not immediately stop a note from playing, it will begin the note's
release phase.

## render

    my $samples = $tsf->render( seconds => 5 );
    my $samples = $tsf->render( samples => 44_100 );

Returns a string of 16-bit, little endian sound samples using TinySoundFont of
the specified length. The result can be unpacked using `unpack("s<*")` or you
can call ["render\_unpack"](#render_unpack) function to get an array instead. This will return
the exact number of samples requested; calling ["render"](#render) 5 times at 1 second
each is identical to calling ["render"](#render) once for 5 seconds.

- seconds

    This sets how many samples to generate in terms of seconds. The default is 1
    second. If both ["seconds"](#seconds) and ["samples"](#samples) are given, seconds will be used.

- samples

    This sets how many samples to generate in terms of seconds. The default is
    ["SAMPLE\_RATE"](#sample_rate). If both ["seconds"](#seconds) and ["samples"](#samples) are given, seconds will
    be used.

## render\_unpack

    my @samples = $tsf->render_unpack(%options);

Returns an array of of 16-bit sound samples using TinySoundFont. All of the
options are identical to ["render"](#render).

## db\_to\_vol

    my $volume = $tsf->db_to_vol(-10);

Convert from dB to a floating point volume. The dB is expressed as a number
between -100 and 0, and will map logarithmically to 0.0 to 1.0.

# Terminology

The SoundFount terminology can get confusing at times, so I've included a quick
reference to help make sense of these.

- Terminology used directly in Audio::TinySoundFont

    To be able to use Audio::TinySoundFont, you will need to know a couple simple
    terms. They are likely easy to infer, but they are here so that their meaning
    is explicit instead of implicit.

    - SoundFont

        Sometimes referred to as SoundFont2, this is a file format designed to store and
        exchange synthesizer information. This includes the audio samples to create
        the audio, how to generate and modify the samples to sound as expected, and
        generally how to produce audio that the SoundFont creator wanted.

    - Preset

        A Preset is the largest usable building block in a SoundFont and generally
        represents a single instrument. If you've ever sat down to an electric keyboard
        and selected "Electric Guitar", "Violin" or "Synth", you are selecting the
        equivalent of a SoundFont preset.

- Terminology used when talking about SoundFonts

    Reading the entire SoundFont2 specification can be daunting. In short, it is a
    [RIFF](https://en.wikipedia.org/wiki/Resource_Interchange_File_Format) file that
    primarily holds 9 types of data, or "sub-chunks". All of this data ultimately
    describes a Preset that is used in constructing audio by TinySoundFont. It is
    not required to understand these to use Audio::TinySoundFont and will not be
    used in the rest of the documentation.

    - PHDR/PBAG (Preset)

        These are the two sections that describe a Preset. A PHDR record describes a
        Preset like the name, preset number and bank number. The PBAG contains what are
        called Preset Zones which lists the generators and modulators to use for a
        specific preset given a range of notes and velocity. One of those generators is
        which Instrument to use, which has its own set of generators and modulators.

    - INST/IBAT (Instrument)

        An instrument is very similar in structure to a Preset, but provides a layer
        of indirection between the raw samples and the presets. A single Instrument
        can be used in multiple Presets, for instance a single Guitar Instrument can be
        used for a regular guitar as well one with extra reverb.
        These two sections serve the same function as the Preset sections. A INST
        describes an Instrument and the PBAG contains Instrument Zones which lists the
        generators and modulators to use for a given range of notes and velocities. One
        of the generators is the sample to use for this Instrument.

    - PGEN
    - PMOD
    - IGEN
    - IMOD

        Each preset and instrument is composed of one or more generators and modulators.
        The PGEN and PMOD sections are used to construct the Preset Zones, likewise the
        IGEN and IMOD sections are used to construct Instrument Zones.
        They describe a single aspect of how to construct the audio samples, for
        instance adding a low-pass filter or reverb. Some are only available for a
        preset, like what Instrument to use, and some are only available to Instruments
        like what Sample to use.

        Note: TinySoundFont does not currently process modulators.

    - SHDR (Samples)

        This section describes the actual audio samples to be used, including a name,
        the length, the original pitch, pitch correction, and looping configuration.
        The actual audio samples are stored in a different RIFF chunk, but this
        contains the references into that chunk about where to find the data.

# AUTHOR

Jon Gentle <cpan@atrodo.org>

# COPYRIGHT

Copyright 2020- Jon Gentle

# LICENSE

This is free software. You may redistribute copies of it under the terms of the Artistic License 2 as published by The Perl Foundation.

# SEE ALSO

[Audio::TinySoundFont::Preset](https://metacpan.org/pod/Audio%3A%3ATinySoundFont%3A%3APreset), [Audio::TinySoundFont::Builder](https://metacpan.org/pod/Audio%3A%3ATinySoundFont%3A%3ABuilder), [TinySoundFont](https://github.com/schellingb/TinySoundFont)


Changes
Revision history for Audio-TinySoundFont

{{$NEXT}}

0.11  2021-01-11 23:39:33 EST
        * Fix building on threaded perls

0.10  2021-01-11 00:10:30 EST
        * Initial release


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.041.
use strict;
use warnings;

use 5.014000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Blah blah blah",
  "AUTHOR" => "Jon Gentle <cpan\@atrodo.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Audio-TinySoundFont",
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.014000",
  "NAME" => "Audio::TinySoundFont",
  "PREREQ_PM" => {},
  "TEST_REQUIRES" => {
    "Test::More" => "0.96"
  },
  "VERSION" => "0.01",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Test::More" => "0.96"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


mk_tiny_sf2.pl
my $TEN0 = "\0\0\0\0\0\0\0\0\0\0";
print ''
  . "RIFF\x{40}\x{02}\x{00}\x{00}sfbk"
  . "LIST\x{2e}\x{00}\x{00}\x{00}INFOifil\x{04}\x{00}\x{00}\x{00}\x{02}\x{00}\x{00}\x{00}isng\x{08}\x{00}\x{00}\x{00}EMU8000\x{00}INAM\x{06}\x{00}\x{00}\x{00}tiny \x{00}"
  . "LIST\x{70}\x{00}\x{00}\x{00}sdtasmpl\x{64}\x{00}\x{00}\x{00}\x{56}\x{00}\x{77}\x{03}\x{1f}\x{07}\x{93}\x{0a}\x{2b}\x{0e}\x{a9}\x{11}\x{3a}\x{15}\x{bd}\x{18}\x{49}\x{1c}\x{cc}\x{1f}\x{49}\x{23}\x{f9}\x{26}\x{2e}\x{2a}\x{47}\x{2e}\x{fa}\x{30}\x{96}\x{35}\x{f2}\x{37}\x{7e}\x{3c}\x{97}\x{3f}\x{6c}\x{42}\x{7e}\x{48}\x{cf}"
  . "\x{46}\x{56}\x{53}\x{64}\x{48}\x{4a}\x{64}\x{a3}\x{27}\x{f1}\x{a3}\x{3b}\x{af}\x{3b}\x{b3}\x{09}\x{b3}\x{86}\x{bb}\x{06}\x{ba}\x{02}\x{c2}\x{05}\x{c2}\x{0f}\x{c8}\x{06}\x{ca}\x{60}\x{ce}\x{9f}\x{d1}\x{23}\x{d5}\x{d5}\x{d8}\x{2d}\x{dc}\x{dd}\x{df}\x{4c}\x{e3}\x{dd}\x{e6}\x{5b}\x{ea}\x{f2}\x{ed}\x{69}\x{f1}\x{08}\x{f5}\x{76}\x{f8}\x{20}\x{fc}"
  . "LIST\x{86}\x{01}\x{00}\x{00}pdta"
  . "phdr\x{72}\x{00}\x{00}\x{00}"
  . "$TEN0$TEN0\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}"
  . "test\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}$TEN0\x{00}\x{00}\x{00}\x{00}\x{01}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}"
  . "$TEN0$TEN0\x{ff}\x{00}\x{ff}\x{00}\x{02}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}"
  . "pbag\x{0C}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{01}\x{00}\x{00}\x{00}\x{02}\x{00}\x{00}\x{00}pmod\x{0a}$TEN0\x{00}\x{00}\x{00}pgen\x{0C}\x{00}\x{00}\x{00}\x{29}\x{00}\x{00}\x{00}\x{29}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}"
  . "inst\x{2c}$TEN0$TEN0\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}$TEN0\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{01}\x{00}"
  . "ibag\x{08}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{02}\x{00}\x{00}\x{00}imod\x{0a}$TEN0\x{00}\x{00}\x{00}"
  . "igen\x{0c}\x{00}\x{00}\x{00}\x{36}\x{00}\x{01}\x{00}\x{35}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}"
  . "shdr\x{5c}$TEN0$TEN0\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{32}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{31}\x{00}\x{00}\x{00}\x{22}\x{56}\x{00}\x{00}\x{3c}\x{00}\x{00}\x{00}\x{01}$TEN0$TEN0$TEN0$TEN0\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}\x{00}"


lib/Audio/TinySoundFont/Preset.pm
package Audio::TinySoundFont::Preset;

use v5.14;
use warnings;
our $VERSION = '0.01';

use autodie;
use Carp;
use Try::Tiny;
use Moo;
use Types::Standard qw/Int Str InstanceOf/;

has soundfont => (
  is       => 'ro',
  isa      => InstanceOf ['Audio::TinySoundFont'],
  required => 1,
);

has index => (
  is       => 'ro',
  isa      => Int,
  required => 1,
);

has name => (
  is      => 'ro',
  isa     => Str,
  lazy    => 1,
  builder => sub
  {
    my $self = shift;
    $self->soundfont->_tsf->get_presetname( $self->index ) // '';
  },
);

sub render
{
  my $self = shift;
  my %args = @_;

  my $tsf = $self->soundfont->_tsf;

  croak "Cannot render a preset when TinySoundFont is active"
      if $tsf->active_voices;

  my $SR      = $tsf->SAMPLE_RATE;
  my $seconds = $args{seconds} // 0;
  my $samples = ( $seconds * $SR ) || $args{samples} // $SR;
  my $note    = $args{note} // 60;
  my $vel     = $args{vel} // 0.5;
  my $vol     = $args{volume} // $self->soundfont->db_to_vol( $args{db} );

  my $old_vol;
  if ( defined $vol )
  {
    $old_vol = $self->soundfont->volume;
    $self->soundfont->volume($vol);
  }

  my $vel_msg = qq{Velocity of "$vel" should be between 0 and 1};
  if ( $vel < 0 )
  {
    carp qq{$vel_msg, setting to 0};
    $vel = 0;
  }

  if ( $vel > 1 )
  {
    carp qq{$vel_msg, setting to 1};
    $vel = 1;
  }

  my $note_msg = qq{Note "$note" should be between 0 and 127};
  if ( $note < 0 )
  {
    carp qq{$note_msg, setting to 0};
    $note = 0;
  }

  if ( $note > 127 )
  {
    carp qq{$note_msg, setting to 127};
    $note = 127;
  }

  $tsf->note_on( $self->index, $note, $vel );

  my $result = $tsf->render($samples);
  $tsf->note_off( $self->index, $note );

  my $cleanup_samples = 4096;
  for ( 1 .. 256 )
  {
    last
        if !$tsf->active_voices;
    $result .= $tsf->render($cleanup_samples);
  }

  if ( defined $old_vol )
  {
    $self->soundfont->volume($old_vol);
  }

  return $result;
}

sub render_unpack
{
  my $self = shift;

  return unpack( 's<*', $self->render(@_) );
}

1;
__END__

=encoding utf-8

=head1 NAME

Audio::TinySoundFont::Preset - SoundFont Preset Represntation

=head1 SYNOPSIS

  use Audio::TinySoundFont;
  my $preset = $tsf->preset('Clarinet');
  my $audio  = $preset->render(seconds => 5, note => 59, vel => 0.7, volume => .3);

=head1 DESCRIPTION

Audio::TinySoundFont::Preset is a Preset, or musical sample in the SoundFont
nomenclature. It is the largest usable building block in a SoundFont and
generally represents a single instrument.

=head1 CONSTRUCTOR

Audio::TinySoundFont::Preset is not constructed manually, instead you get an
instance from the L<preset|Audio::TinySoundFont/preset> or L<preset_index|Audio::TinySoundFont/preset_index> methods of L<Audio::TinySoundFont>.

=head1 METHODS

=head2 soundfont

The L<Audio::TinySoundFont> object that this Preset object was created from.

=head2 index

The index of this preset in the SoundFont file.

=head2 name

The name of the preset. For example, it could be "Clarinet", "Accordion",
"Standard Drums" or "Sine Wave".

=head2 render

  my $samples = $preset->render(%options);

Returns a string of 16-bit, little endian sound samples for this
preset using TinySoundFont. The result can be unpacked using C<unpack("s<*")>
or you can call L</render_unpack> function to get an array instead. It accepts
several options that allow you to customize how the sound should be generated.

This method cannot be used if the L</soundfont> is currently playing anything
and will croak if that is the case.

=head3 options

=over

=item seconds

This sets how long the preset should be played for, from Attack to Release.
It defaults to 1 second. If both L</seconds> and L</samples> are given,
seconds will be used.

NOTE: The sample is not constrained to the number of seconds given, it is the
minimum number of seconds that will be returned. The sound may continue to
play after these many seconds, and often will, because of the decay phase
of the SoundFont.

=item samples

This is an alternative way to set how long the preset should be played for,
from Attack to Release, measure in 16-bit samples. It defaults to 1 second
worth of samples. If both L</seconds> and L</samples> are given, seconds will
be used.

NOTE: The sample is not constrained to the number of samples given, it is the
minimum number of samples that will be returned. The sound may continue to
play after these many samples, and often will, because of the decay phase
of the SoundFont.

=item note

This is the MIDI note to be played with 60 meaning middle C, which is the
default. It is an integer between 0 and 127.

=item vel

The MIDI velocity of the note played, expressed as a float between 0 and 1,
the default is 0.5. Generally a velocity causes a louder noise to be played,
but the SoundFont can use it modify other aspects of the sound. This is
separate from L</volume> which controls the general volume of the resulting
sample.

=item volume

The general volume of the sample, expressed as a float between 0 and 1.
If both L</volume> and L</db> are given, volume will be used.

=item db

The volume of the sample, expressed in as a dB float between -100 and 0, with
-100 being equivalent to silent and 0 being as loud as TinySoundFont can go.
If both L</volume> and L</db> are given, volume will be used.

=back

=head2 render_unpack

  my @samples = $preset->render_unpack(%options);

Returns an array of of 16-bit sound samples for this preset using TinySoundFont.
All of the options are identical to L</render>.

=head1 AUTHOR

Jon Gentle E<lt>cpan@atrodo.orgE<gt>

=head1 COPYRIGHT

Copyright 2020- Jon Gentle

=head1 LICENSE

This is free software. You may redistribute copies of it under the terms of the Artistic License 2 as published by The Perl Foundation.

=head1 SEE ALSO

L<Audio::TinySoundFont>, L<TinySoundFont|https://github.com/schellingb/TinySoundFont>.

=cut


lib/Audio/TinySoundFont.pm
package Audio::TinySoundFont;

use v5.14;
use warnings;
our $VERSION = '0.01';

use autodie;
use Carp;
use Try::Tiny;
use Scalar::Util qw/blessed/;

use Moo;
use Types::Standard qw/ArrayRef HashRef GlobRef Str Int Num InstanceOf/;

use Audio::TinySoundFont::XS;
use Audio::TinySoundFont::Preset;
use Audio::TinySoundFont::Builder;

has _tsf => (
  is       => 'ro',
  isa      => InstanceOf ['Audio::TinySoundFont::XS'],
  required => 1,
);

has volume => (
  is      => 'rw',
  isa     => Num,
  default => 0.3,
  trigger => sub { my $self = shift; $self->_tsf->set_volume(shift) },
);

has preset_count => (
  is  => 'lazy',
  isa => Int,
);

has presets => (
  is  => 'lazy',
  isa => HashRef,
);

*SAMPLE_RATE = \&Audio::TinySoundFont::XS::SAMPLE_RATE;

my $XS        = 'Audio::TinySoundFont::XS';
my %ref_build = (
  '' => sub
  {
    my $file = shift;
    croak qq{File "$file" doesn't exist}
        if !-e $file;
    return try { $XS->load_file($file) } catch { croak $_ };
  },
  SCALAR => sub
  {
    my $str = shift;
    open my $glob, '<', $str;
    return try { $XS->load_fh($glob) } catch { croak $_ };
  },
  GLOB => sub
  {
    my $fh = shift;
    return try { $XS->load_fh($fh) } catch { croak $_ };
  },
);

sub BUILDARGS
{
  my $class = shift;
  my $file  = shift;
  my $args  = Moo::Object::BUILDARGS( $class, @_ );

  my $build_fn = $ref_build{ ref $file };
  croak "Cannot load soundfont file, unknown ref: " . ref($file)
      if !defined $build_fn;
  my $tsf = $build_fn->($file);
  $args->{volume} = 0.3;

  $args->{_tsf} = $tsf;

  return $args;
}

sub _build_preset_count
{
  my $self = shift;

  return $self->_tsf->presetcount;
}

sub _build_presets
{
  my $self = shift;

  my %result;
  foreach my $i ( 0 .. $self->preset_count )
  {
    my $name     = $self->_tsf->get_presetname($i) // '';
    my $n        = '';
    my $conflict = 1;
    while ( exists $result{"$name$n"} )
    {
      $conflict++;
      $n = "_$conflict";
    }
    $name = "$name$n";
    $result{$name} = Audio::TinySoundFont::Preset->new(
      soundfont => $self,
      index     => $i,
    );
  }

  return \%result;
}

sub preset
{
  my $self = shift;
  my $name = shift;

  my $preset = $self->presets->{$name};

  croak qq{Could not find preset "$name"}
      if !defined $preset;

  return $preset;
}

sub preset_index
{
  my $self  = shift;
  my $index = shift;

  croak qq{Could not find preset "$index"}
      if $index >= $self->preset_count;

  return Audio::TinySoundFont::Preset->new(
    soundfont => $self,
    index     => $index,
  );
}

sub new_builder
{
  my $self   = shift;
  my @script = @_;

  if ( @script == 1 && ref $script[0] eq 'ARRAY' )
  {
    @script = @{ $script[0] };
  }

  return Audio::TinySoundFont::Builder->new(
    soundfont   => $self,
    play_script => \@script,
  );
}

sub active_voices
{
  my $self = shift;
  return $self->_tsf->active_voices;
}

sub is_active
{
  my $self = shift;
  return !!$self->_tsf->active_voices;
}

sub note_on
{
  my $self   = shift;
  my $preset = shift // croak "Preset is required for note_on";
  my $note   = shift // 60;
  my $vel    = shift // 0.5;

  if ( !blessed $preset )
  {
    $preset = $self->preset($preset);
  }

  ( InstanceOf ['Audio::TinySoundFont::Preset'] )->($preset);

  $self->_tsf->note_on( $preset->index, $note, $vel );
  return;
}

sub note_off
{
  my $self   = shift;
  my $preset = shift // croak "Preset is required for note_off";
  my $note   = shift // 60;

  if ( !blessed $preset )
  {
    $preset = $self->preset($preset);
  }

  ( InstanceOf ['Audio::TinySoundFont::Preset'] )->($preset);

  $self->_tsf->note_off( $preset->index, $note );
  return;
}

sub render
{
  my $self = shift;
  my %args = @_;

  my $tsf = $self->_tsf;

  my $SR      = $tsf->SAMPLE_RATE;
  my $seconds = $args{seconds} // 0;
  my $samples = ( $seconds * $SR ) || $args{samples} // $SR;

  return $tsf->render($samples);
}

sub render_unpack
{
  my $self = shift;

  return unpack( 's<*', $self->render(@_) );
}

sub db_to_vol
{
  my $self = shift;
  my $db   = shift;

  return
      if !defined $db;

  # Volume is a float 0.0-1.0, db is in dB -100..0, so adjust it to a float
  $db
      = $db > 0    ? 0
      : $db < -100 ? -100
      :              $db;
  return 10**( $db / 20 );
}

1;
__END__

=encoding utf-8

=head1 NAME

Audio::TinySoundFont - Interface to TinySoundFont, a "SoundFont2 synthesizer library in a single C/C++ file"

=head1 SYNOPSIS

  use Audio::TinySoundFont;
  my $tsf = Audio::TinySoundFont->new('soundfont.sf2');
  $tsf->note_on( 'Clarinet', 60, 1.0 );
  my $samples = $tsf->render( seconds => 5 );
  my $preset = $tsf->preset('Clarinet');
  $tsf->note_off( $preset, 60 );
  $samples .= $tsf->render( seconds => 5 );
  
  # Using the Preset object
  my $preset = $tsf->preset('Clarinet');
  my $samples = $preset->render(
    seconds => 5,
    note => 60,
    vel => 0.7,
    volume => .3,
  );
  
  # Using the Builder object
  my $builder = $tsf->new_builder(
    [
      {
        preset => 'Clarinet',
        note   => 59,
        at     => 0,
        for    => 2,
      },
    ]
  );
  $builder->add(
    [
      {
        preset     => $preset,
        note       => 60,
        at         => 44100,
        for        => 44100 * 2,
        in_seconds => 0,
      },
    ]
  );
  my $samples = $builder->render;

=head1 DESCRIPTION

Audio::TinySoundFont is a wrapper around L<TinySoundFont|https://github.com/schellingb/TinySoundFont>,
a "SoundFont2 synthesizer library in a single C/C++ file". This allows you to
load a SoundFont file and synthesize samples from it.

=head1 CONSTRUCTOR

=head2 new

  my $tsf = Audio::TinySoundFont->new($soundfont_file, %attributes)

Construct a new Audio::TinySoundFont object using the provided C<$soundfont_file>
file. It can be a filename, a file handle, or a string reference to the contents
of a SoundFont.

=head3 Attributes

=over

=item volume

Set the initial, global volume. This is a floating point number between 0.0 and
1.0. The higher the number, the louder the output samples.

=back

=head1 METHODS

=head2 volume

  my $volume = $tsf->volume;
  $tsf->volume( 0.5 );

Get or set the current global volume. This is a floating point number between
0.0 and 1.0. The higher the number, the louder the output samples.

=head2 preset_count

  my $count = $tsf->preset_count;

The number of presets available in the SoundFont.

=head2 presets

  my %presets = %{ $tsf->presets }

A HashRef of all of the presets in the SoundFont. The key is the name of the
preset and the value is the L<Audio::TinySoundFont::Preset> object for that
preset.

=head2 preset

  my $preset = $tsf->preset('Clarinet');

Get a L<Audio::TinySoundFont::Preset> object by name. This will croak if the
preset name is not found.

=head2 preset_index

  my $preset = $tsf->preset_index($index);

Get an L<Audio::TinySoundFont::Preset> object by index in the SoundFont. This
will croak if the index is out of range. Note, this will return a different
object than L</preset> will, which will return the object from L<presets>.

=head2 SAMPLE_RATE

  my $sample_rate = $tsf->SAMPLE_RATE

Returns the sample rate that TinySoundFont is operating on, expressed as hertz
or samples per second. This is currently static and is 44_100;

=head2 new_builder

  my $builder = $tsf->new_builder

Create a new L<Audio::TinySoundFont::Builder> object. This can be used to
generate a single sample from a script of what notes to play when.

=head2 active_voices

  my $count = $tsf->active_voices;

Returns the number of currently active voices that TinySoundFont is rendering.
Generally speaking, each L</note_on> will make one or more voices active.

=head2 is_active

  my $bool = $tsf->is_active

Returns if TinySoundFont currently has active voices and will output audio
during render.

=head2 note_on

  $tsf->note_on($preset, $note, $velocity);

Turns a note on for a Preset. C<$preset> can either be a
L<Audio::TinySoundFont::Preset> object or the name of a Preset. Both C<$note>
and C<$velocity> are optional. C<$note> is a MIDI note between 0 and 127, with
60 being middle C and the default if it is not given. C<$velocity> is a floating
point number between 0.0 and 1.0 with the default being 0.5.

=head2 note_off

  $tsf->note_off($preset, $note);

Turns a note off for a Preset. C<$preset> can either be a
L<Audio::TinySoundFont::Preset> object or the name of a Preset. C<$note> is
optional and is the same MIDI note given on L</note_on>. The default is 60.
This will not immediately stop a note from playing, it will begin the note's
release phase.

=head2 render

  my $samples = $tsf->render( seconds => 5 );
  my $samples = $tsf->render( samples => 44_100 );

Returns a string of 16-bit, little endian sound samples using TinySoundFont of
the specified length. The result can be unpacked using C<unpack("s<*")> or you
can call L</render_unpack> function to get an array instead. This will return
the exact number of samples requested; calling L</render> 5 times at 1 second
each is identical to calling L</render> once for 5 seconds.

=over

=item seconds

This sets how many samples to generate in terms of seconds. The default is 1
second. If both L</seconds> and L</samples> are given, seconds will be used.

=item samples

This sets how many samples to generate in terms of seconds. The default is
L</SAMPLE_RATE>. If both L</seconds> and L</samples> are given, seconds will
be used.

=back

=head2 render_unpack

  my @samples = $tsf->render_unpack(%options);

Returns an array of of 16-bit sound samples using TinySoundFont. All of the
options are identical to L</render>.

=head2 db_to_vol

  my $volume = $tsf->db_to_vol(-10);

Convert from dB to a floating point volume. The dB is expressed as a number
between -100 and 0, and will map logarithmically to 0.0 to 1.0.

=head1 Terminology

The SoundFount terminology can get confusing at times, so I've included a quick
reference to help make sense of these.

=over

=item Terminology used directly in Audio::TinySoundFont

To be able to use Audio::TinySoundFont, you will need to know a couple simple
terms. They are likely easy to infer, but they are here so that their meaning
is explicit instead of implicit.

=over

=item SoundFont

Sometimes referred to as SoundFont2, this is a file format designed to store and
exchange synthesizer information. This includes the audio samples to create
the audio, how to generate and modify the samples to sound as expected, and
generally how to produce audio that the SoundFont creator wanted.

=item Preset

A Preset is the largest usable building block in a SoundFont and generally
represents a single instrument. If you've ever sat down to an electric keyboard
and selected "Electric Guitar", "Violin" or "Synth", you are selecting the
equivalent of a SoundFont preset.

=back

=item Terminology used when talking about SoundFonts

Reading the entire SoundFont2 specification can be daunting. In short, it is a
L<RIFF|https://en.wikipedia.org/wiki/Resource_Interchange_File_Format> file that
primarily holds 9 types of data, or "sub-chunks". All of this data ultimately
describes a Preset that is used in constructing audio by TinySoundFont. It is
not required to understand these to use Audio::TinySoundFont and will not be
used in the rest of the documentation.

=over

=item PHDR/PBAG (Preset)

These are the two sections that describe a Preset. A PHDR record describes a
Preset like the name, preset number and bank number. The PBAG contains what are
called Preset Zones which lists the generators and modulators to use for a
specific preset given a range of notes and velocity. One of those generators is
which Instrument to use, which has its own set of generators and modulators.

=item INST/IBAT (Instrument)

An instrument is very similar in structure to a Preset, but provides a layer
of indirection between the raw samples and the presets. A single Instrument
can be used in multiple Presets, for instance a single Guitar Instrument can be
used for a regular guitar as well one with extra reverb.
These two sections serve the same function as the Preset sections. A INST
describes an Instrument and the PBAG contains Instrument Zones which lists the
generators and modulators to use for a given range of notes and velocities. One
of the generators is the sample to use for this Instrument.

=item PGEN

=item PMOD

=item IGEN

=item IMOD

Each preset and instrument is composed of one or more generators and modulators.
The PGEN and PMOD sections are used to construct the Preset Zones, likewise the
IGEN and IMOD sections are used to construct Instrument Zones.
They describe a single aspect of how to construct the audio samples, for
instance adding a low-pass filter or reverb. Some are only available for a
preset, like what Instrument to use, and some are only available to Instruments
like what Sample to use.

Note: TinySoundFont does not currently process modulators.

=item SHDR (Samples)

This section describes the actual audio samples to be used, including a name,
the length, the original pitch, pitch correction, and looping configuration.
The actual audio samples are stored in a different RIFF chunk, but this
contains the references into that chunk about where to find the data.

=back

=back

=head1 AUTHOR

Jon Gentle E<lt>cpan@atrodo.orgE<gt>

=head1 COPYRIGHT

Copyright 2020- Jon Gentle

=head1 LICENSE

This is free software. You may redistribute copies of it under the terms of the Artistic License 2 as published by The Perl Foundation.

=head1 SEE ALSO

L<Audio::TinySoundFont::Preset>, L<Audio::TinySoundFont::Builder>, L<TinySoundFont|https://github.com/schellingb/TinySoundFont>

=cut
